Add First Pass Tracking Functionality

emdavis02 opened the pull request:
## PR Description

The `sleap-mot` repo currently supports tracking pose-estimated videos either:
1. From scratch (no prior identity information), or
2. With initial tracklets produced from **first-pass tracking**.

First-pass tracking uses **visible features** or **external signals** to assign partial identities (tracklets) to animals when those features are available. For example:
- Visual features like **fur color** can distinguish animals when they're unobstructed.
- **RFID pings** can be matched to individuals when a signal is detected.

Previously, these feature-based tracking methods were implemented in scattered Jupyter notebooks. The goal of this PR is to **incorporate them directly into the `sleap-mot` repository**, enabling it to function as a **complete identity tracking pipeline**, from raw pose data to globally consistent tracks.

---

## Key Changes

- **Introduced a general `FeatureTracker` base class**  
  This abstract base class defines shared functionality and interfaces for all feature-based tracking approaches. It includes:
  - Utility functions for tracklet management
  - A common structure for integrating with the broader tracking pipeline
  - An **abstract method** `track()` that must be implemented differently in each subclass depending on the feature type

**Implemented subclasses:**
  - **`RFIDFeatureTracker`**  
    This subclass demonstrates how specific feature-based methods (in this case, RFID-based) can be cleanly integrated using the inherited structure and shared utilities from `FeatureTracker`.
  - **`FurColorFeatureTracker`**  

- Test functions will be written to test the functionality of these new classes in tandem with the tracker class (second step tracking)

---

## Outcome

This PR formalizes and integrates feature-based first-pass tracking into the core `sleap-mot` package. It lays the groundwork for a modular, extensible identity tracking system, where multiple feature types (e.g., color, RFID, embeddings) can be easily added and composed.

emdavis02 committed:
initial classes and RFID tracker functions

emdavis02 committed:
succesfully implemented RFIDFeatureTracker

emdavis02 committed:
black

emdavis02 committed:
docstring

emdavis02 committed:
docstring

emdavis02 committed:
FurColor tracking class implemented

emdavis02 committed:
black

talmo commented on sleap_mot/feature_tracker.py:
```diff
@@ -188,12 +192,20 @@ def extract_tracking_data(self, labels):
                 # Initialize a dictionary for this row with 'frame_idx' and 'track'
                 row = {"frame_idx": frame_data.frame_idx, "track": inst.track.name}
                 for point in inst.points:
```
Could do something like:

```py
cols = [*[f"{node_name}.x", f"{node_name}.y", f"{node_name}.score"] for node_name in skeleton.node_names]

pd.DataFrame(inst.numpy(scores=True).reshape(-1), columns=cols)
```

emdavis02 committed:
updating point syntax

emdavis02 committed:
tail tracking

emdavis02 committed:
Add shapely dependency and enhance docstrings in feature_tracker and tracker classes

- updated environment
- Expanded docstrings in the FeatureTracker class to include detailed argument descriptions.
- Implemented force matching for untracked instances in the Tracker class, including methods for resolving track IDs and finding closest global track instances.
- Improved handling of untracked instances and global tracks during tracking process.

emdavis02 committed:
pydocstring corrections

emdavis02 committed:
pydocstring fixes

emdavis02 committed:
improving error handling for user experience

emdavis02 committed:
black

emdavis02 committed:
pydocstring
